Add support for Samsung Galaxy A71

WiiPlayer2 opened the pull request:
This includes device-tree and generated panel driver.

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2023, Danila Tikhonov <danila@jiaxyga.com>
```
Add your copyright too if you want

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+};
+
+/* Limit CPU clock to 2.2 GHz */
+ &cpu6_opp14 {
```
Drop it

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	chassis-type = "handset";
+
+	qcom,msm-id = <QCOM_ID_SM7150 0>;
+	qcom,board-id = <34 0>, /* Rev0.0 */
```
msm-id and board-id are not required when using header v2. If there is no reason to use another header version, then drop it

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+		wifi0 = &wifi;
+	};
+
+	battery: battery {
```
Are you sure Samsung uses qcom pmic for fg?

WiiPlayer2 replied:
`qcom,qgauge@4800` is enabled in the downstream device tree if that's what you mean.

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+		qcom,pmic-id = "a";
+
+		vreg_s4a_1p1: smps4 {
+			regulator-min-microvolt = <824000>;
```
Add the regulator-name property to each regulator

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	status = "okay";
+};
+
+// If enabled device crashes. It might be directly this device or a dependent
```
Well then fix it or disable it

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+
+&gpu {
+	zap-shader {
+		firmware-name = "qcom/sm7150/a715f/a615_zap.mbn";
```
All firmware paths: qcom/sm7150/samsung/a715f/*

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	status = "okay";
+
+	panel@0 {
+		compatible = "mdss,s6e3fa9-amb667um06";
```
```
compatible = "samsung,s6e3fa9-amb667um06-a715f", "samsung,s6e3fa9-amb667um06";
```

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+};
+
+&mdss_dsi0_phy {
+			// vdda-0p9-supply = <0x1a1>;
```
?

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	};
+};
+
+&pm6150_qgauge {
```
Same question as for battery node

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	status = "okay";
+};
+
+&qup_spi1_spi {
```
I don't see spi1 used on this board

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+};
+
+&tlmm {
+	gpio-reserved-ranges = <0 4>, <59 4>, <83 2>;
```
Add a comment describing what these pins are used for

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	sde_te_active: sde-te-state {
+		pins = "gpio10";
+		function = "mdp_vsync";
+		drive-strength = <0x02>;
```
Use decimal values wherever possible:
drive-strength = <2>;

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+		vddxo-supply = <&vreg_l1c_1p8>;
+		vddrf-supply = <&vreg_l2c_1p3>;
+		vddch0-supply = <&vreg_l10c_3p3>;
+		max-speed = <3200000>; // TODO: copied; still need to verify
```
I can assure you that 3200000 is the correct value, please delete the comment

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	status = "okay";
+
+	// NOTE: needs this after every reboot: 'sudo btmgmt --index 0 public-addr 00:11:22:33:44:55'
+	// see: https://matrix.to/#/!IeSVTOYowpUldDcgJO:matrix.org/$SrXjQ6LfPBMPk-qDWwhQDPtVsgR-8gJTnSDKPATnVNU?via=matrix.org&via=tchncs.de&via=nitro.chat
```
Why is this even here?

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+};
+
+// TODO: wifi is disabled until it doesn't crash the modem remote processor anymore.
+&wifi {
```
Please see dt-bindings. Otherwise, drop it

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	vdd-3.3-ch1-supply = <&vreg_l11c_3p3>; // NOTE: downstream does not define any supply for ch1
+	// firmware-name = "qcom/sm7150/a715f/wlanmdsp.mbn";
+	status = "disabled";
+};
```
Fix any warnings that dt-bindings check gives you:
https://www.linaro.org/blog/tips-and-tricks-for-validating-devicetree-sources-with-the-devicetree-schema/

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+	regulators-0 {
+		compatible = "qcom,pm6150-rpmh-regulators";
+		qcom,pmic-id = "a";
+
```
What about the regulators supply map?

JIaxyga commented on arch/arm64/configs/sm7150.config:
```diff
@@ -41,6 +41,9 @@ CONFIG_SND_SOC_TAS2562=m
 CONFIG_DRM_PANEL_SAMSUNG_AMS667YR01=y
 
+# Samsung Galaxy A71 (A715F)
+CONFIG_DRM_PANEL_S6E3FA9_AMB667UM06_FHD=y
```
This should definitely be a separate commit

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+	return 0;
+}
+
+// TODO: Check if /sys/class/backlight/.../actual_brightness actually returns
```
Well then, check?

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+	struct device *dev = &dsi->dev;
+	const struct backlight_properties props = {
+		.type = BACKLIGHT_RAW,
+		.brightness = 425,
```
This is an AMOLED panel, probably you need to set at least half of the default brightness?

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -669,34 +674,76 @@
 &tlmm {
 	gpio-reserved-ranges = <0 4>, <59 4>, <83 2>;
 
+	gpio-line-names =
```
This is fundamentally wrong, the array of pins does not correspond to the number of gpio pins in the SoC

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -533,10 +533,15 @@
+		pinctrl-1 = <&sde_te_suspend &sde_dsi_suspend &sde_esd_suspend &sde_ub_det_suspend>;
 
 		vddio-supply = <&vreg_l13a_1p8>;
 		vdd3p3-supply = <&vreg_l18a_2p8>;
```
I don't see that your panel driver supports regulators

JIaxyga commented on arch/arm64/boot/dts/qcom/Makefile:
```diff
@@ -257,6 +257,7 @@ dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-surya-huaxing.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-surya-tianma.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-sweet.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-tucana.dtb
+dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-samsung-a715f.dtb
```
Merge into the dts commit

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -11,6 +11,7 @@
 #include <dt-bindings/leds/common.h>
```
Merge into the dts commit

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -155,8 +155,6 @@ static int s6e3fa9_amb667um06_bl_update_status(struct backlight_device *bl)
 	return 0;
```
Merge into the panel commit

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,788 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Look at how commits are formatted with other devices (preferably those that are already in upstream). I mean commit name and message

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
Same remark on commit formatting as DTS commit

JIaxyga commented on drivers/gpu/drm/panel/Kconfig:
```diff
@@ -1062,4 +1062,14 @@ config DRM_PANEL_XINPENG_XPP055C272
 	  XPP055C272 controller for 720x1280 LCD panels with MIPI/RGB/SPI
 	  system interfaces.
+
+config DRM_PANEL_S6E3FA9_AMB667UM06_FHD
```
Place the panel in alphabetical order, not at the end

JIaxyga commented on drivers/gpu/drm/panel/Makefile:
```diff
@@ -108,3 +108,4 @@ obj-$(CONFIG_DRM_PANEL_VISIONOX_VTDR6130) += panel-visionox-vtdr6130.o
 obj-$(CONFIG_DRM_PANEL_VISIONOX_R66451) += panel-visionox-r66451.o
 obj-$(CONFIG_DRM_PANEL_WIDECHIPS_WS2401) += panel-widechips-ws2401.o
 obj-$(CONFIG_DRM_PANEL_XINPENG_XPP055C272) += panel-xinpeng-xpp055c272.o
+obj-$(CONFIG_DRM_PANEL_S6E3FA9_AMB667UM06_FHD) += panel-s6e3fa9-amb667um06.o
```
alphabetical order and why you use FHD in the config name?

JIaxyga commented on drivers/gpu/drm/panel/Kconfig:
```diff
@@ -1062,4 +1062,14 @@ config DRM_PANEL_XINPENG_XPP055C272
+	depends on BACKLIGHT_CLASS_DEVICE
+	help
+	  Say Y here if you want to enable support for the
+	  S6E3FA9_AMB667UM06_FHD controller for 1080x2400 LCD panels with MIPI/RGB/SPI
```
wtf? spi?

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+}
+
+static const struct of_device_id s6e3fa9_amb667um06_of_match[] = {
+	{ .compatible = "mdss,s6e3fa9-amb667um06" }, // FIXME
```
So fix

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+};
+module_mipi_dsi_driver(s6e3fa9_amb667um06_driver);
+
+MODULE_AUTHOR("linux-mdss-dsi-panel-driver-generator <fix@me>"); // FIXME
```
So fix

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+
+	drm_panel_add(&ctx->panel);
+
+	ret = mipi_dsi_attach(dsi);
```
use devm_mipi_dsi_attach

JIaxyga commented on drivers/gpu/drm/panel/panel-s6e3fa9-amb667um06.c:
```diff
@@ -0,0 +1,268 @@
+	return container_of(panel, struct s6e3fa9_amb667um06, panel);
+}
+
+static void s6e3fa9_amb667um06_reset(struct s6e3fa9_amb667um06 *ctx)
```
Are you sure about this sequence? I ask this regarding your comment in dts to the panel

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -533,10 +533,15 @@
+		In the downstream kernel pin 11 of &tlmm is used for resetting the display.
+		If it's actually set here then the display will turn off as soon as the pin is set to 0 and apparently it won't turn on again.
+		*/
+		reset-gpios = <&tlmm 9 GPIO_ACTIVE_HIGH>;
```
Since it doesn't work the best solution is to remove it. Why use wrong values?

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -669,34 +674,76 @@
 		bias-disable = <0x00>;
 	};
 
+	sde_esd_suspend: sde-esd-state {
```
Same pin configuration as sde_dsi suspend. Extend the pin list of sde_dsi_suspend instead of a separate node

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,875 @@
+};
+
+&dispcc {
+	status = "okay";
```
I don't see dispcc being disabled:
https://github.com/sm7150-mainline/linux/blob/v6.14/arch/arm64/boot/dts/qcom/sm7150.dtsi#L3973

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -0,0 +1,875 @@
+
+&i2c0 {
+	clock-frequency = <100000>;
+	status = "okay";
```
Maybe you could at least leave a comment on what devices are connected? How can we be sure that this bus is used at all? Otherwise just don't enable i2cX until needed. I'm almost certainly sure that one of these buses is not used on your board, and the pins from the bus were reused. This is why you have a crashes when you enable gpi1

JIaxyga commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -44,7 +44,7 @@
 };
 
 / {
-	model = "Samsung A71";
```
Merge with DTS commit

JIaxyga commented on drivers/gpu/drm/panel/Kconfig:
```diff
@@ -1062,4 +1062,14 @@ config DRM_PANEL_XINPENG_XPP055C272
 	  Say Y here if you want to enable support for the Xinpeng
```
Write dt-bindings for the panel driver

JIaxyga reviewed (changes requested):
This is just a quick review. Please fix all of this first of all

Gelbpunkt commented on arch/arm64/boot/dts/qcom/sm7150-samsung-a715f.dts:
```diff
@@ -606,10 +606,10 @@
 		 * If it's not set at all the display won't show anything because reset-gpios is required by the driver.
 		*/
-		reset-gpios = <&tlmm 9 GPIO_ACTIVE_HIGH>; // TODO: Check this again or fix in driver
+		reset-gpios = <&tlmm 11 GPIO_ACTIVE_HIGH>;
```
your reset gpio is almost certainly active low and that should fix all your problems

WiiPlayer2 committed:
arm64: dts: qcom: sm7150: Add device-tree for Samsung Galaxy A71

Add device tree for the Samsung Galaxy A71 (a715f) smartphone. This
device is based on Snapdragon 730G (sm7150) SoC.

Co-developed-by: Danila Tikhonov <danila@jiaxyga.com>

WiiPlayer2 commented:
This now only includes the device tree adition
